smtk-import/smartrak.cpp

```diff
@@ -698,49 +670,44 @@ static std::string smtk_locate_buddy(MdbHandle *mdb, char *dive_idx, char *buddi
  * The "tag" parameter is used to mark if we want this table to be imported
  * into tags or into notes.
  */
-static void smtk_parse_relations(MdbHandle *mdb, struct dive *dive, char *dive_idx, const char *table_name, const char *rel_table_name, char *list[], bool tag)
+static void smtk_parse_relations(MdbHandle *mdb, struct dive *dive, char *dive_idx, const char *table_name, const char *rel_table_name, const std::vector<std::string> &list, bool tag)
 {
 	std::string tmp;
-	struct types_list *diverel_head, *d_runner;
-
-	diverel_head = smtk_index_list(mdb, rel_table_name, dive_idx);
-	if (!diverel_head)
-		return;
 
 	/* Get the text associated with the relations */
-	for (d_runner = diverel_head; d_runner; d_runner = d_runner->next) {
-		if (!list[d_runner->idx - 1])
+	std::vector<int> diverel_list = smtk_index_list(mdb, rel_table_name, dive_idx);
+	for (int idx: diverel_list) {
+		const std::string str = get(list, idx - 1);
+		if (str.empty())
 			continue;
 		if (tag)
-			taglist_add_tag(&dive->tag_list, list[d_runner->idx - 1]);
+			taglist_add_tag(&dive->tag_list, str.c_str());
 		else
-			concat(tmp, ", ", std::string(list[d_runner->idx - 1]));
-		if (strstr(list[d_runner->idx - 1], "SCR"))
+			concat(tmp, ", ", str);
+		if (str.find("SCR") != std::string::npos)
 			dive->dc.divemode = PSCR;
-		else if (strstr(list[d_runner->idx -1], "CCR"))
+		else if (str.find("CCR") != std::string::npos)
 			dive->dc.divemode = CCR;
 	}
 	if (!tmp.empty())
 		concat(dive->notes, "\n", format_string_std("Smartrak %s: %s", table_name, tmp.c_str()));
-	smtk_list_free(diverel_head);
 }
 
 /*
  * Add data from tables related in Dives table which are not directly supported
  * in Subsurface. Write them as tags or dive notes by setting true or false the
  * boolean parameter "tag".
  */
-static void smtk_parse_other(struct dive *dive, char *list[], const char *data_name, char *idx, bool tag)
+static void smtk_parse_other(struct dive *dive, const std::vector<std::string> &list, const char *data_name, char *idx, bool tag)
 {
        int i = atoi(idx) - 1;
-       char *str = NULL;
 
```

Comment:
I would add a check here for i value >= 0.  If a table is read wrongly we can get idx=0 and thus i < 0.

Just 
`
        if (i < 0)
              return;  // idx should be 1 at least
`
Works fine.